// create-amplify/lib/amplify_project_creator.js
import { logger } from './logger.js';
const LEARN_MORE_USAGE_DATA_TRACKING_LINK = `https://docs.amplify.aws/gen2/reference/telemetry`;
/**
 *
 */
export class AmplifyProjectCreator {
    packageManagerController;
    projectRootValidator;
    initialProjectFileGenerator;
    npmInitializedEnsurer;
    gitIgnoreInitializer;
    projectRoot;
    defaultDevPackages = [
        '@aws-amplify/backend',
        '@aws-amplify/backend-cli',
        'aws-cdk@^2',
        'aws-cdk-lib@^2',
        'constructs@^10.0.0',
        'typescript@^5.0.0',
    ];
    defaultProdPackages = ['aws-amplify'];
    /**
     * Orchestrator for the create-amplify workflow.
     * Delegates out to other classes that handle parts of the getting started experience
     */
    constructor(packageManagerController, projectRootValidator, initialProjectFileGenerator, npmInitializedEnsurer, gitIgnoreInitializer, projectRoot) {
        this.packageManagerController = packageManagerController;
        this.projectRootValidator = projectRootValidator;
        this.initialProjectFileGenerator = initialProjectFileGenerator;
        this.npmInitializedEnsurer = npmInitializedEnsurer;
        this.gitIgnoreInitializer = gitIgnoreInitializer;
        this.projectRoot = projectRoot;
    }
    /**
     * Executes the create-amplify workflow
     */
    create = async () => {
        logger.debug(`Validating current state of target directory...`);
        await this.projectRootValidator.validate();
        await this.npmInitializedEnsurer.ensureInitialized();
        await logger.indicateProgress(`Installing required dependencies`, async () => {
            await this.packageManagerController.installDependencies(this.defaultProdPackages, 'prod');
            await this.packageManagerController.installDependencies(this.defaultDevPackages, 'dev');
        });
        await logger.indicateProgress(`Creating template files`, async () => {
            await this.gitIgnoreInitializer.ensureInitialized();
            await this.initialProjectFileGenerator.generateInitialProjectFiles();
        });
        logger.log('Successfully created a new project!');
        const cdCommand = process.cwd() === this.projectRoot
            ? ''
            : `cd .${this.projectRoot.replace(process.cwd(), '')}; `;
        logger.log(`Welcome to AWS Amplify! 
Run \`npx amplify help\` for a list of available commands. 
Get started by running \`${cdCommand}npx amplify sandbox\`.`);
        logger.log(`Amplify (Gen 2) collects anonymous telemetry data about general usage of the CLI.

Participation is optional, and you may opt-out by using \`amplify configure telemetry disable\`.

To learn more about telemetry, visit ${LEARN_MORE_USAGE_DATA_TRACKING_LINK}`);
    };
}
